crates/katana/rpc/rpc/src/torii.rs

```diff
@@ -0,0 +1,131 @@
+
+            let provider = this.sequencer.backend.blockchain.provider();
+            let latest_block_number = this.sequencer.block_number().map_err(ToriiApiError::from)?;
+
+            if cursor.block_number > latest_block_number + 1 {
+                return Err(ToriiApiError::BlockNotFound.into());
+            }
+
+            if latest_block_number >= cursor.block_number {
+                for block_number in cursor.block_number..=latest_block_number {
+                    let tx_range = BlockProvider::block_body_indices(
+                        provider,
+                        BlockHashOrNumber::Num(block_number),
+                    )
+                    .map_err(ToriiApiError::from)?
+                    .ok_or(ToriiApiError::BlockNotFound)?;
+
+                    let mut block_transactions = provider
+                        .transaction_in_range(Range::from(tx_range))
+                        .map_err(ToriiApiError::from)?
+                        .into_iter()
+                        .map(|tx| tx.clone().into())
+                        .collect::<Vec<_>>();
+
+                    // If the block_number is the cursor block, slice the transactions from the txn
+                    // offset
+                    if block_number == cursor.block_number {
+                        block_transactions = block_transactions
+                            .iter()
+                            .skip(cursor.transaction_index as usize)
+                            .cloned()
+                            .collect();
+                    }
+
+                    transactions.extend(block_transactions);
+                }
+            }
+
+            if let Some(pending_state) = this.sequencer.pending_state() {
+                let pending_transactions = pending_state
+                    .executed_txs
+                    .read()
+                    .iter()
+                    .map(|(tx, _)| tx.clone().into())
+                    .collect::<Vec<_>>();
+
+                // If cursor is in the pending block
+                if cursor.block_number == latest_block_number + 1 {
+                    if cursor.transaction_index as usize > pending_transactions.len() {
+                        return Err(ToriiApiError::TransactionOutOfBounds.into());
+                    }
+
+                    let mut pending_transactions = pending_transactions
+                        .iter()
+                        .skip(cursor.transaction_index as usize)
+                        .cloned()
+                        .collect::<Vec<_>>();
+
+                    // If there are no transactions after the index in the pending block
+                    if pending_transactions.is_empty() {
+                        // Wait for a new transaction to be added to the pool
+                        let mut rx = this.sequencer.pool.add_listener();
+                        pending_transactions = match futures::executor::block_on(rx.next()) {
```

Comment:
@kariy this yields when a new txn is received in the pool but we actually want it after the txn has been executed, so it is available on the pending state. What's the best approach to doing that?